logstash-core/lib/logstash/runner.rb

```diff
@@ -102,30 +106,35 @@ def execute
 
     return start_shell(@ruby_shell, binding) if @ruby_shell
 
-    if @config_string.nil? && @config_path.nil?
+    @agent = create_agent
+
+    if (config_string.nil? || config_string.empty?) && config_path.nil?
       fail(I18n.t("logstash.runner.missing-configuration"))
     end
 
-    @agent.logger = @logger
-
-    config_string = format_config(@config_path, @config_string)
-
-    @agent.add_pipeline("base", config_string, :filter_workers => filter_workers)
+    config_str = LogStash::Config::Loader.format_config(config_path, config_string)
 
     if config_test?
-      puts "Configuration OK"
+      config_error = @agent.config_valid?(config_str)
+      if config_error
+        @logger.fatal I18n.t("logstash.error", :error => config_error)
+        return 1
+      else
+        @logger.terminal "Configuration OK"
+      end
     else
+      @agent.add_pipeline("base", config_str, :filter_workers => filter_workers)
       task = Stud::Task.new { @agent.execute }
       return task.wait
     end
 
+  rescue MissingAgentError => e
+    return 1
```

Comment:
No, because before the exception is raised a `@logger.fatal` is called. It is confusing however. I'll try to rework the flow.